// @aragon/connect/dist/cjs/connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
const connect_core_1 = require("@aragon/connect-core");
const connect_ethereum_1 = __importDefault(require("@aragon/connect-ethereum"));
const connect_thegraph_1 = __importDefault(require("@aragon/connect-thegraph"));
const constants_1 = require("./constants");
function ipfsResolverFromUrlTemplate(urlTemplate) {
    return function ipfsResolver(cid, path) {
        const url = urlTemplate.replace(/\{cid\}/, cid);
        if (!path) {
            return url.replace(/\{path\}/, '');
        }
        if (!path.startsWith('/')) {
            path = `/${path}`;
        }
        return url.replace(/\{path\}/, path);
    };
}
function getIpfsResolver(ipfs) {
    if (typeof ipfs === 'function') {
        return ipfs;
    }
    return ipfsResolverFromUrlTemplate(ipfs || constants_1.DEFAULT_IPFS_URL);
}
function normalizeConnectorConfig(connector) {
    if (Array.isArray(connector)) {
        return [connector[0], connector[1] || {}];
    }
    if (typeof connector === 'string') {
        return [connector, {}];
    }
    return null;
}
function getConnector(connector, network) {
    const normalizedConfig = normalizeConnectorConfig(connector);
    if (normalizedConfig === null) {
        return connector;
    }
    const [name, config] = normalizedConfig;
    if (!config.network) {
        config.network = network;
    }
    if (name === 'json') {
        return new connect_core_1.ConnectorJson(config);
    }
    if (name === 'thegraph') {
        return new connect_thegraph_1.default(config);
    }
    if (name === 'ethereum') {
        return new connect_ethereum_1.default(config);
    }
    throw new Error(`Unsupported connector name: ${name}`);
}
function getEthersProvider(ethereumProvider, network) {
    // Ethers compatibility: ethereum => homestead
    if (network.name === 'ethereum' && network.chainId === 1) {
        network = { ...network, name: 'homestead' };
    }
    if (ethereumProvider) {
        try {
            return new ethers_1.ethers.providers.Web3Provider(ethereumProvider, network);
        }
        catch (err) {
            console.error('Invalid provider:', ethereumProvider);
            throw err;
        }
    }
    if (network.chainId === 100) {
        return new ethers_1.ethers.providers.WebSocketProvider(constants_1.XDAI_WSS_ENDPOINT, network);
    }
    return ethers_1.ethers.getDefaultProvider(network);
}
async function resolveAddress(ethersProvider, location) {
    const address = ethers_1.ethers.utils.isAddress(location)
        ? location
        : await ethersProvider.resolveName(location);
    if (!ethers_1.ethers.utils.isAddress(address)) {
        throw new Error('Please provide a valid address or ENS domain.');
    }
    return address;
}
async function connect(location, connector, { actAs, ethereum: ethereumProvider, ipfs, network, verbose, } = {}) {
    var _a;
    const _network = connect_core_1.toNetwork(network !== null && network !== void 0 ? network : 'ethereum');
    const ethersProvider = getEthersProvider(ethereumProvider, _network);
    const orgConnector = getConnector(connector, _network);
    const orgAddress = await resolveAddress(ethersProvider, location);
    const connectionContext = {
        actAs: actAs || null,
        ethereumProvider: ethereumProvider || null,
        ethersProvider,
        ipfs: getIpfsResolver(ipfs),
        network: _network,
        orgAddress,
        orgConnector,
        orgLocation: location,
        verbose: verbose !== null && verbose !== void 0 ? verbose : false,
    };
    await ((_a = orgConnector.connect) === null || _a === void 0 ? void 0 : _a.call(orgConnector, connectionContext));
    return new connect_core_1.Organization(connectionContext);
}
exports.default = connect;
