// @aragon/connect-core/dist/cjs/entities/Organization.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TransactionIntent_1 = __importDefault(require("../transactions/TransactionIntent"));
const misc_1 = require("../utils/misc");
function normalizeAppFilters(filters) {
    var _a;
    if (!filters) {
        return {};
    }
    if (typeof filters === 'string') {
        return filters.startsWith('0x')
            ? { address: [filters] }
            : { name: [filters] };
    }
    if (Array.isArray(filters)) {
        return ((_a = filters[0]) === null || _a === void 0 ? void 0 : _a.startsWith('0x')) ? { address: filters }
            : { name: filters };
    }
    if (filters.address) {
        return { address: misc_1.toArrayEntry(filters.address) };
    }
    if (filters.name) {
        return { name: misc_1.toArrayEntry(filters.name) };
    }
    return {};
}
class Organization {
    constructor(connection) {
        this.connection = connection;
    }
    get location() {
        return this.connection.orgLocation;
    }
    get address() {
        return this.connection.orgAddress;
    }
    get _connection() {
        return this.connection;
    }
    ///////// APPS ///////////
    async app(filters) {
        return this.connection.orgConnector.appForOrg(this, normalizeAppFilters(filters));
    }
    async apps(filters) {
        return this.connection.orgConnector.appsForOrg(this, normalizeAppFilters(filters));
    }
    onApp(filtersOrCallback, callback) {
        const filters = (callback ? filtersOrCallback : null);
        const _callback = (callback || filtersOrCallback);
        return this.connection.orgConnector.onAppForOrg(this, normalizeAppFilters(filters), _callback);
    }
    onApps(filtersOrCallback, callback) {
        const filters = (callback ? filtersOrCallback : null);
        const _callback = (callback || filtersOrCallback);
        return this.connection.orgConnector.onAppsForOrg(this, normalizeAppFilters(filters), _callback);
    }
    ///////// PERMISSIONS ///////////
    async permissions() {
        return this.connection.orgConnector.permissionsForOrg(this);
    }
    onPermissions(callback) {
        return this.connection.orgConnector.onPermissionsForOrg(this, callback);
    }
    ///////// INTENTS ///////////
    appIntent(appAddress, functionName, functionArgs) {
        return new TransactionIntent_1.default({ contractAddress: appAddress, functionName, functionArgs }, this, this.connection.ethersProvider);
    }
}
exports.default = Organization;
